### PR TITLE
Create servers array in openapi schema

### PR DESCRIPTION
**Pull request by alisihab**

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java [hunk 1]
```diff
@@ -286,6 +288,22 @@ public String getOperationId() {
...
 
+	@IbisDoc({"12", "Comma separated list of parameter names defined in the header. Do not use followings 'Accept', 'Content-Type' or 'Authorization'", ""})
+	public void setHeaderParams(String headerParams) {
+		this.headerParams = headerParams;
+	}
+	public String getHeaderParams() {
+		return headerParams;
+	}
+
+	@IbisDoc({"13", "Comma separated list of parameter names defined in the cookie.", ""})
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java:299]: I think 'defined in the cookie' should be: 'passed as cookie'

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java [hunk 2]
```diff
@@ -286,6 +288,22 @@ public String getOperationId() {
 		return operationId;
 	}
 
+	@IbisDoc({"12", "Comma separated list of parameter names defined in the header. Do not use followings 'Accept', 'Content-Type' or 'Authorization'", ""})
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java:291]: I think 'defined in the header' should be: 'passed as http header'.
Why can the specified list not be used?

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java [hunk 3]
```diff
@@ -205,11 +207,41 @@ protected JsonObject generateOpenApiJsonSchema(Collection<ApiDispatchConfig> cli
...
 	}
 
+	private JsonArrayBuilder mapServers(HttpServletRequest request) {
+		JsonArrayBuilder serversArray = Json.createArrayBuilder();
+		String protocol = request.isSecure() ? "https://" : "http://";
+		String suffix = "/api";
+		String env = AppConstants.getInstance().getString("dtap.stage", "LOC");
+		int port = request.getServerPort();
+		// Get load balancer url if exists
+		String loadBalancerUrl = AppConstants.getInstance().getProperty("load.balancer.url", null);
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java:224]: Dots are not spaces, the dot's divide categories into sub-categories; `jdbc.dateFormat` generic database related format and `jdbc.datasource.default` datasource related. I think loadbalancer should either be 1 word or have a underscore (load_balancer)

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java [hunk 4]
```diff
@@ -205,11 +207,41 @@ protected JsonObject generateOpenApiJsonSchema(Collection<ApiDispatchConfig> cli
...
+		if(StringUtils.isNotEmpty(loadBalancerUrl)) {
+			serversArray.add(Json.createObjectBuilder().add("url", loadBalancerUrl + suffix).add("description", "load balancer for " + env + " server"));
+		}
+		// Get details from properties
+		String hostname = AppConstants.getInstance().getString("hostname", null);
+		if(StringUtils.isNotEmpty(hostname)) {
+			String webContextPath = AppConstants.getInstance().getString("web.contextpath", null);
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java:231]: I think it's a bad idea to make up this many new properties

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java [hunk 5]
```diff
@@ -205,11 +207,41 @@ protected JsonObject generateOpenApiJsonSchema(Collection<ApiDispatchConfig> cli
...
+		String suffix = "/api";
+		String env = AppConstants.getInstance().getString("dtap.stage", "LOC");
+		int port = request.getServerPort();
+		// Get load balancer url if exists
+		String loadBalancerUrl = AppConstants.getInstance().getProperty("load.balancer.url", null);
+		if(StringUtils.isNotEmpty(loadBalancerUrl)) {
+			serversArray.add(Json.createObjectBuilder().add("url", loadBalancerUrl + suffix).add("description", "load balancer for " + env + " server"));
+		}
+		// Get details from properties
+		String hostname = AppConstants.getInstance().getString("hostname", null);
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java:229]: this is a very bad idea... the server request path should be used, or a property, nothing else.

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java [hunk 6]
```diff
@@ -475,6 +475,24 @@ else if(segment.startsWith("{") && segment.endsWith("}")) {
...
 				}
+				if(StringUtils.isNotEmpty(listener.getHeaderParams())) {
+					String params[] = listener.getHeaderParams().split(",");
+					for (String headerParam : params) {
+						String param = request.getHeader(headerParam);
+						if(param != null) {
+							messageContext.put(headerParam, param);
+						}
+					}
+				}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java:486]: headers are already mapped to the messageContext

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java [hunk 7]
```diff
@@ -475,6 +475,24 @@ else if(segment.startsWith("{") && segment.endsWith("}")) {
...
+				if(StringUtils.isNotEmpty(listener.getCookieParams())) {
+					String params[] = listener.getCookieParams().split(",");
+					for (String cookieParam : params) {
+						Cookie cookie = CookieUtil.getCookie(request, cookieParam);
+						if(cookie != null) {
+							messageContext.put(cookieParam, cookie.getValue());
+						}
+					}
+				}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java:495]: We don't do anything with cookies, if you are going to implement this, at least put a restriction on it. Don't just pass all cookies but only those set in the listener.

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java [hunk 8]
```diff
@@ -286,6 +288,22 @@ public String getOperationId() {
 		return operationId;
 	}
 
+	@IbisDoc({"12", "Comma separated list of parameters passed as http header.", ""})
+	public void setHeaderParams(String headerParams) {
+		this.headerParams = headerParams;
+	}
+	public String getHeaderParams() {
+		return headerParams;
+	}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiListener.java:297]: If you are adding this I expect a validation check and only add the headers stated in this setter to be added to the messageContext

File: test/src/test/testtool/Validators/Schema/openapi.json [hunk 9]
```diff
@@ -5,6 +5,20 @@
...
+            "url":"https://loadbalancer.url",
+            "description":"load balancer for LOC server"
+        },
+        {
+            "url":"http://hostname:8085/iaf-test/api",
+            "description":"LOC server hostname"
+        },
+        {
+            "url":"http://localhost:8085/iaf-test/api",
+            "description":"LOC server"
```
Comment (nielsm5) [inline test/src/test/testtool/Validators/Schema/openapi.json:19]: It should not return the same information twice

File: core/src/main/resources/AppConstants.properties [hunk 10]
```diff
@@ -423,4 +423,7 @@ console.idle.timeout=0
 console.dateFormat=yyyy-MM-dd HH:mm:ss
 
 #Feedback URL to be used
-console.feedbackURL=https://ibissource.org/iaf/feedback/
\ No newline at end of file
+console.feedbackURL=https://ibissource.org/iaf/feedback/
+
+#Load balancer url(this will be used to create server object in openapi schema)
+load.balancer.url=
```
Comment (nielsm5) [inline core/src/main/resources/AppConstants.properties:429]: You've added the load.balancer.url but not `hostname` and `web.contextpath`?

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java [hunk 11]
```diff
@@ -205,11 +207,33 @@ protected JsonObject generateOpenApiJsonSchema(Collection<ApiDispatchConfig> cli
...
+		components.add("schemas", schemas);
+		root.add("components", components);
 
 		return root.build();
 	}
 
+	private JsonArrayBuilder mapServers(HttpServletRequest request) {
+		JsonArrayBuilder serversArray = Json.createArrayBuilder();
+		String protocol = request.isSecure() ? "https://" : "http://";
+		String suffix = "/api";
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/http/rest/ApiServiceDispatcher.java:220]: Could you retrieve the suffix from this property `servlet.ApiListenerServlet.urlMapping` and fall back to the default (api) when not set?

**Review by nielsm5 — CHANGES_REQUESTED** [review]

**Review by nielsm5 — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java (hunk 7 above)
Comment (gvanbrakel) [reply core/src/main/java/nl/nn/adapterframework/http/rest/ApiListenerServlet.java:487]: I agree with Niels. I think it is better we comment this code out for now.

File: core/src/test/resources/OpenApi/cookieParams.json [hunk 12]
```diff
@@ -0,0 +1,83 @@
+{
+    "openapi":"3.0.0",
+    "info":{
+        "title":"Sample API",
```
Comment (gvanbrakel) [inline core/src/test/resources/OpenApi/cookieParams.json:4]: Title should be name of the adapter

File: core/src/test/resources/OpenApi/cookieParams.json [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        "description":"OpenApi document auto-generated by Frank!Framework for test",
+        "version":"0.1.9"
```
Comment (gvanbrakel) [inline core/src/test/resources/OpenApi/cookieParams.json:6]: Version should be the configuration version, where available. Not just `0.1.9`

File: test/src/test/testtool/Validators/Schema/openapi_post.json [hunk 14]
```diff
@@ -5,6 +5,20 @@
...
+        },
+        {
+            "url":"http://hostname:8085/iaf-test/api",
+            "description":"LOC server hostname"
+        },
+        {
+            "url":"http://localhost:8085/iaf-test/api",
+            "description":"LOC server"
+        }
+    ],
```
Comment (gvanbrakel) [inline test/src/test/testtool/Validators/Schema/openapi_post.json:21]: We should not expose internal urls when not necessary.
So if 'loadBalancer.url' is specified, not other hostname should be returned

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — APPROVED** [review]

**Pull request merged**